src/query/src/part_sort.rs

```diff
@@ -659,20 +776,41 @@ impl PartSortStream {
         if this_range.num_rows() != 0 {
             self.push_buffer(this_range)?;
         }
-        // mark end of current PartitionRange
-        let sorted_batch = self.sort_buffer();
-        // step to next proper PartitionRange
+
+        // Step to next proper PartitionRange
         self.cur_part_idx += 1;
 
-        // If we've processed all partitions, discard remaining data
+        // If we've processed all partitions, sort and output
         if self.cur_part_idx >= self.partition_ranges.len() {
             // assert there is no data beyond the last partition range (remaining is empty).
-            // it would be acceptable even if it happens, because `remaining_range` will be discarded anyway.
             debug_assert!(remaining_range.num_rows() == 0);
 
+            // Sort and output the final group
+            let sorted_batch = self.sort_buffer();
             return sorted_batch.map(|x| if x.num_rows() == 0 { None } else { Some(x) });
         }
 
+        // Check if we're still in the same group
+        if self.is_in_current_group(self.cur_part_idx) {
+            // Same group - don't sort yet, keep collecting
+            let next_sort_column = sort_column.slice(idx, batch.num_rows() - idx);
+            if self.try_find_next_range(&next_sort_column)?.is_some() {
+                // remaining batch still contains data that exceeds the current partition range
+                self.evaluating_batch = Some(remaining_range);
+            } else {
+                // remaining batch is within the current partition range
+                if remaining_range.num_rows() != 0 {
+                    self.push_buffer(remaining_range)?;
+                }
+            }
```

Comment:
Is this possible? We have already called try_find_next_range before.

Reply:
in case a batch contains data from two partition ranges. this is possible after batch collapsing.